_posts/2020/2020-01-20-be-suspicious-of-join-tables.md

```diff
@@ -0,0 +1,112 @@
+---
+title: "Be Suspicious of Join Tables"
+description: "They're probably hiding a concept you haven’t named yet"
+layout: article
+category: ruby
+image:
+  base: '2020/be-suspicious-of-join-tables'
+  alt: Welding
+  credit:  Max LaRochelle
+  source: "https://unsplash.com/photos/QzP1GcDOSC8"
+
+---
+
+We often have to represent many-to-many relationships between models in our applications. Rails provides a method in the migrations to generate a table in your database to support this. You can see the documentation in the [Rails guide for ActiveRecord migrations](https://guides.rubyonrails.org/active_record_migrations.html#creating-a-join-table).
+
+However these “pure” join tables often obscure a useful concept in your application that might be better represented as a named model.
+
+
+## Instead of...
+
+...using a join table:
+
+#### Migration
+
+```ruby
+create_join_table :user, :organisation
+```
+
+#### `app/models/user.rb`
+
+```ruby
+class User < ApplicationRecord
+  has_and_belongs_to_many :organisations
+end
+```
+
+#### `app/models/organisation.rb`
+
+```ruby
+class Organisation < ApplicationRecord
+  has_and_belongs_to_many :users
+end
+```
+
+## Use
+
+...a real model, by naming the concept that the join table is hiding.
+
+#### Migration
+
+```ruby
+create_table :memberships do |t|
+  t.references :user
+  t.references :organisation
+end
+```
+
+#### `app/models/membership.rb`
+
+```ruby
+class Membership < ApplicationRecord
+  belongs_to :user
+  belongs_to :organisation
+end
+```
+
+#### `app/models/user.rb`
+
+```ruby
+class User < ApplicationRecord
+  has_many :memberships
+  has_many :organisations, through: :memberships
+end
+```
+
+#### `app/models/organisation.rb`
+
+```ruby
+class Organisation < ApplicationRecord
+  has_many :memberships
+  has_many :users, through: :memberships
+end
+```
+
+
+## Why?
+
+Almost without fail, whenever a join model sits for any length of time in an application it begins to acquire behaviour. It’s nearly always worth working out and having a first pass at naming the concept that the join model represents.
+
+With a “pure” join table there is no way to add functionality to this unnamed concept. The lack of a place to put this extension means you may accidentally attach functionality to one of the joined models.
+
+For example it might be putting a `role` attribute on the `User`  in this example, making later refactoring harder.
+
+
+## Why not?
+
+There’s extra manual work to be done when you can’t simply create “has and belongs to”-style joins, as you have to create the joining model yourself.
```

Comment:
```suggestion
There’s extra manual work to be done as you have to create the joining model yourself.
```

Reply:
I mean...you can create them, but you're _choosing_ not to. 

Maybe: There's extra manual work to be done because since you're not using the built-in functionality for "has and belongs to"-style joins, you have to create the joining model yourself...